Week 9: Routes for Creating, Updating and Deleting Data

ltphongssvn opened the pull request:
## Week 9 Assignment Resubmission

This pull request implements all requirements for Lesson 9. This is a resubmission to replace PR #1 which could not be reopened due to the missing main branch that has now been created.

### Question 1: Course Form Dropdowns
- Replaced text inputs with dropdown selectors for coding class and trimester
- Used collection_select helper to display user-friendly names while submitting IDs
- Implemented display_name method in Trimester model for better dropdown labels

### Question 2: Filtered Submission Dropdowns
- Populated submission form with filtered data from parent course
- Display only lessons belonging to the specific course
- Display only students enrolled in the course

### Question 3: Trimester Edit/Update Functionality
- Implemented edit form for updating trimester application deadlines
- Added comprehensive validation for date inputs
- Returns appropriate HTTP status codes (400 for bad request, 404 for not found)
- All required RSpec tests passing

### Changes Made:
- Modified routes configuration for RESTful resources
- Enhanced controllers with proper data loading and validation
- Updated views with improved form controls
- Added model methods for better user experience

**Note:** This replaces the accidentally closed PR #1 with identical Week 9 implementation.

ltphongssvn committed:
Implement Week 9 CRUD operations with enhanced form controls

- Add dropdown selectors to course creation form for better UX
- Implement filtered dropdowns in submission form showing only relevant students and lessons
- Create full edit/update functionality for trimester application deadlines
- Add date validation and error handling for trimester updates
- Implement display_name method in Trimester model for user-friendly dropdown labels
- Configure RESTful routes for courses, submissions, and trimesters
- Handle nested submission creation under course context

inlibra-cpa commented on app/controllers/courses_controller.rb:
Using `params.expect` (not a Rails API) will raise and bypass strong params(same for `set_course`).

Here's the standard solution example:
```
def course_params
  params.require(:course).permit(:coding_class_id, :trimester_id, :max_enrollment)
end
```

inlibra-cpa commented on app/controllers/trimesters_controller.rb:
`params[:trimester][:application_deadline]` will `NoMethodError` if `:trimester` is missing. You might want to use `params.dig(:trimester, :application_deadline)` in checks.

inlibra-cpa commented on app/models/trimester.rb:
The assignment asked for space-separated ("Fall 2025"). Update unless your tests explicitly require the hyphen.

inlibra-cpa reviewed (changes requested):
Hi @ltphongssvn ,
Week 9 asks for (1) an edit-page spec that asserts the "application deadline" label, and (2) PUT specs for happy path, missing deadline → 400, invalid date → 400, and missing ID → 404. Please add them.
Please update `params.expect` → `require/permit`, remove the duplicate courses routes, harden the trimester parameter checks, align `display_name` with the spec, and add the Week-9 request specs. Then this PR can be merged.

ltphongssvn committed:
Address Week 9 code review feedback from @inlibra-cpa

- Replace params.expect with standard params.require().permit() in courses_controller
- Fix params access in set_course to use conventional params[:id]
- Update trimesters_controller to use params.dig() for safe nested parameter access
- Change Trimester#display_name format from hyphen to space separator (Fall 2025)
- Add comprehensive request specs for trimesters including:
  * Edit page test verifying application deadline label
  * Happy path test for successful updates
  * Error handling tests for missing deadline (400)
  * Error handling tests for invalid date format (400)
  * Test for non-existent trimester ID (404)

All requested RSpec tests now passing (5/5 examples, 0 failures)

ltphongssvn commented:
@inlibra-cpa Thank you for your thorough code review. I've addressed all your feedback in commit b5ca7e1. Here's what I've updated:

## Controller Changes
- **courses_controller.rb**: Replaced `params.expect` with the standard `params.require(:course).permit(...)` pattern in the `course_params` method
- **courses_controller.rb**: Updated `set_course` method to use conventional `params[:id]` instead of `params.expect(:id)`
- **trimesters_controller.rb**: Fixed the parameter access to use `params.dig(:trimester, :application_deadline)` for safe nested hash navigation, preventing NoMethodError when :trimester key is missing

## Model Changes
- **trimester.rb**: Corrected the `display_name` method to use space separator as specified - now returns "Fall 2025" instead of "Fall-2025"

## Test Coverage
Added comprehensive request specs in `spec/requests/trimesters_spec.rb` with all the tests you requested:
- Edit page spec that asserts the "application deadline" label is present
- PUT request happy path test for successful updates
- PUT request test for missing deadline returning 400 Bad Request
- PUT request test for invalid date format returning 400 Bad Request
- PUT request test for non-existent ID returning 404 Not Found

All 5 RSpec tests are passing successfully (confirmed with `bundle exec rspec spec/requests/trimesters_spec.rb`).

The routes configuration was already clean with no duplicates - `resources :courses` and `resources :trimesters` are each defined only once.

These changes improve the code's robustness, follow Rails conventions more closely, and ensure proper error handling for edge cases. Please let me know if you need any additional changes or have questions about the implementation.

shaun-technovation commented on app/views/courses/_form.html.erb:
👍

shaun-technovation commented on app/views/courses/_form.html.erb:
👍

shaun-technovation commented on app/views/trimesters/edit.html.erb:
👍

shaun-technovation commented on spec/requests/trimesters_spec.rb:
Nice specs!

shaun-technovation reviewed (approved):
Awesome, great work here! Also, nice overview/summaries in all of your PRs! 💯 👏

ltphongssvn commented:
Thank you, I appreciate it!
